### ISSUE TITLE
stb_image.h: Cannot get stbi_load_from_memory to work

### ISSUE THREAD

**Issue by Blacklands** [nothings/stb#332]
Hi,

Let me just state that I have been using stb_image for a while in various little (OpenGL-related) C++ projects, and love it! Thanks for creating this awesome loader! :)

I'm now experiencing my first problem with the library, which I presume is NOT a bug, but simply me incorrectly using it:
As the title says, I'm totally unable to load any image directly from memory, using stbi_load_from_memory. Previously, I only loaded images from the hard disk, which always worked fine.

I have a memory stream (an std::istream) containing the image, which I then load into a vector\<unsigned char\>, and then I try to load that image with the above function (using the data buffer of the vector). So it goes:
[some place] -> istream -> vector[buffer] -> stbi_load_from_memory
However, I always only get a null pointer back.

I've compiled an example below:
http://pastebin.com/0X5zgaQT

I've been using this image for testing: 
![bird](https://cloud.githubusercontent.com/assets/4943543/16546032/56d559fa-413d-11e6-9838-c92e419f50d6.jpg)

I included some commented-out stuff into the code for checking that my stream indeed contains the correct image data, and that loading the image from the hard disk with stb_image actually works.
I've been trying this on Windows (7), using Visual Studio 2015. 

What am I doing wrong here? Am I misunderstanding the parameters of stbi_read_from_memory? Or even the entire purpose of the function? Or does it somehow incorrectly recognize the data buffer of my vector (I was under the impression that the buffer is just a normal unsigned char* array/memory block)?

I know that this is most likely not a bug, given that there do not seem to be any other issues regarding loading images from memory from what I could find.
I wasn't sure if I should ask this here or on StackOverflow, but I figured that here would be a better place (correct me if I'm wrong, please).

Thanks in advance for any help (even if it's just "please ask this somewhere else")!

**Comment by nothings** [comment]
This isn't the place for non-bug discussion, but as long as the thread exists, I might as well answer.

The first step is, what is the result of calling of stbi_failure_reason() after it fails to load?

**Comment by zao** [comment]
Your test case is not legal, you may want to verify that you have indeed properly read the data first.

`main` shall return int, and way more importantly: `std::ios::badbit` is not part of the `std::ios::openmode` enumeration. You're masking in some bit in the mask that may have some very undesired implementation defined meaning, like `std::ios::ate` or something similarly amusing.

The image as downloaded from there works for me, given one fixes the trivial errors in the code.

**Comment by Blacklands** [comment]
@nothings I'm sorry then. I wasn't sure; I'll not open an issue for stuff like this another time. Thanks for deciding to still help!
@zao My fault. At some point in the past I forgot about how main() should even look like. The thing is that the compiler Visual Studio uses actually allows this syntax (void main()), apparently this is some kind of weird extension. I used it once at some point and never questioned it again later when it worked. That's what you get for not double-checking...
About the actual issue: I feel _so_ stupid now. As you see, I'm too blindly trusting the compiler to catch me when I write stupid code. I used streams left and right and managed to get confused by the different types of bits. Makes sense that the compiler wouldn't know about what kinds of bits are used for which streams, and what they do. Now I know that I have to be _much_ more careful about this in the future.
After throwing out the ios::badbit, it indeed works, like you said. Let me go into the shaming corner where I belong now.

I'm sorry that I wasted both your time with something as trivial as this. Thank you again for still helping me! I'm closing this now, then.
